File: mobile/lib/motion.ts
/**
 * Motion presets — five named curves, four durations.
 *
 * Why named instead of inline:
 *   · Every spring-bounce inline is a chance to introduce a different
 *     curve and dilute the system. Naming them forces consistency.
 *   · Premium apps have ONE motion personality. Linear feels like
 *     Linear because every transition reads from the same easing
 *     library. We're doing the same.
 *
 * Use:
 *   · `easeOut`     — default for anything that fades / translates in
 *   · `easeOutSoft` — premium slide-ins, sheet expansions
 *   · `easeInOut`   — symmetric transitions (open/close)
 *   · `spring`      — celebration ONLY (moment animations)
 *   · `linear`      — ticking timers / progress bars
 *
 * Durations:
 *   · `fast`        — 180ms — interactive feedback (press, toggle)
 *   · `base`        — 320ms — most screen-level transitions
 *   · `slow`        — 500ms — sheet open, modal in
 *   · `celebrate`   — 1200ms — moment animations (Strava-style)
 *
 * For Reanimated worklets we re-export the timing helpers tuned to
 * these durations + curves — call sites get `t.base`, `t.slow`, etc.
 */

import { Easing, type WithTimingConfig } from "react-native-reanimated";

import { curves, durations } from "./theme";

export { curves, durations };

/** Reanimated Easing functions matching our named curves. */
export const ease = {
  easeOut: Easing.bezier(...curves.easeOut),
  easeOutSoft: Easing.bezier(...curves.easeOutSoft),
  easeInOut: Easing.bezier(...curves.easeInOut),
  spring: Easing.bezier(...curves.spring),
  linear: Easing.linear,
} as const;

/** Pre-built `withTiming` configs. Reduces boilerplate at call sites. */
export const timing: Record<
  "fast" | "base" | "slow" | "celebrate",
  WithTimingConfig
> = {
  fast: { duration: durations.fast, easing: ease.easeOut },
  base: { duration: durations.base, easing: ease.easeOutSoft },
  slow: { duration: durations.slow, easing: ease.easeOutSoft },
  celebrate: { duration: durations.celebrate, easing: ease.spring },
};

/**
 * Motion (react-native-reanimated) transition presets for use with
 * `entering` and `exiting` props on Animated.View.
 *
 * Example:
 *   <Animated.View entering={FadeInUp.duration(durations.base).easing(ease.easeOutSoft)}>
 *
 * Most callers just pull `transitions.fadeUp.base` etc.
 */

File: ios-native/BuilderHQ/DesignSystem/Motion.swift
/// Motion — named curves + durations, mirroring the RN v4 system.
///
/// Five curves, four durations. Spring is reserved for celebration
/// moments only (state transitions: tender awarded, project published,
/// builder unlocks). Every other animation uses easeOut or easeOutSoft.
///
/// SwiftUI's `Animation` is the surface layer; UIKit-flavoured callers
/// (rare) can use the raw cubic-bezier helpers below.

import SwiftUI

enum Motion {
    // ── Durations ─────────────────────────────────────────────────
    static let fast       = Duration.milliseconds(180)
    static let base       = Duration.milliseconds(320)
    static let slow       = Duration.milliseconds(500)
    static let celebrate  = Duration.milliseconds(1200)

    // ── Named animations ─────────────────────────────────────────

    /// Default — most things (fades, translates).
    static let easeOut: Animation = .timingCurve(0.2, 0.8, 0.2, 1, duration: 0.32)

    /// Premium feel — slide-ins, sheet expansions, hero entries.
    static let easeOutSoft: Animation = .timingCurve(0.16, 1, 0.3, 1, duration: 0.5)

    /// Reversible — open / close, symmetric transitions.
    static let easeInOut: Animation = .timingCurve(0.65, 0, 0.35, 1, duration: 0.32)

    /// Celebration ONLY — moments of achievement. Big damping ratio
    /// so the overshoot reads as joy, not jitter.
    static let spring: Animation = .spring(response: 0.55, dampingFraction: 0.62)

    /// Linear — ticking timers, progress bars.
    static let linear: Animation = .linear(duration: 0.32)
}

/// Reanimated-style transition presets adapted for SwiftUI's
/// `.transition` modifier. Use for FadeInDown-style entries.
extension AnyTransition {
    /// Fade in while sliding up from 12pt below — the v4 row entry.
    static let fadeInUpV4: AnyTransition = .asymmetric(
        insertion: .move(edge: .bottom).combined(with: .opacity),
        removal: .opacity
    )

    /// Spring-scale entry for the Strava-style celebration moment.
    static let momentZoom: AnyTransition = .asymmetric(
        insertion: .scale(scale: 0.4).combined(with: .opacity),
        removal: .opacity
    )
}
